### PR TITLE
feat(Live-Annotation): support Live-Annotation

### PR DESCRIPTION
**Pull request by perfect**
# COMPLETES #< INSERT LINK TO ISSUE >

SDK support the live annotation attendee section
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-405642 

## by making the following changes

image

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request

- [ ] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/annotation/index.ts [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+   * close annotation
+   * @param {object} requestData
+   * @returns {Promise}
+   */
+  public closeAnnotation(requestData: RequestData) {
+    return this.sendAnnotationAction(ANNOTATION_ACTION_TYPE.CLOSED, requestData);
+  }
+
+  /**
+   * cancel approved annotation
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/annotation/index.ts:202]: issue: this comment is not correct

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 2]
```diff
@@ -2272,6 +2282,7 @@ export default class Meeting extends StatelessWebexPlugin {
     this.locusInfo.on(EVENTS.LOCUS_INFO_UPDATE_URL, (payload) => {
       this.members.locusUrlUpdate(payload);
       this.breakouts.locusUrlUpdate(payload);
+      this.annotation.locusUrlUpdate(payload);
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:2285]: issue: There should be tests to check that the locus url and approval url update are called

File: packages/@webex/plugin-meetings/src/meeting/request.ts [hunk 3]
```diff
@@ -679,6 +691,9 @@ export default class MeetingRequest extends StatelessWebexPlugin {
     if (options?.resourceToken) {
       body.resourceToken = options?.resourceToken;
     }
+    if (options?.annotation) {
+      body.annotation = options?.annotation;
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/meeting/request.ts:695]: issue: There should be a test for this

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts [hunk 4]
```diff
@@ -0,0 +1,349 @@
+import MockWebex from '@webex/test-helper-mock-webex';
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:1]: issue: The convention is to have the test file names the same as file name they are testing. This file should be called index.ts

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+      annotationService.register = sinon.stub().resolves({
+        body: {
+          binding: 'binding',
+          webSocketUrl: 'url',
+        },
+      });
+    });
+
+
+    describe('#MessageProcess', () => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:47]: issue: The convention is to use `#` to indicate that this is a test for a specific method/attribute for a class. i.e. you could have `describe('#eventCommandProcessor())`

Or you could still group the tests as you have done but remove the `#` i.e. `describe('message processing')`

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+
+        await annotationService.processStrokeMessage({encryptionKeyUrl: 'encryptionKeyUrl', content: 'content'});
+
+        assert.calledOnceWithExactly(spy, {
+          data: {encryptionKeyUrl: 'encryptionKeyUrl', content: 'decryptedContent'},
+        });
+
+      });
+
+      it('eventDataProcessor', () => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:89]: issue: as with the eventCommandProcessor you need to test that this function can handle the case when e doesn't have a data attribute (or when e itself is undefined)

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+        assert.calledOnceWithExactly(annotationService.webex.internal.llm.socket.send, sendObject);
+      });
+
+    });
+
+
+    describe('#Locus API collection', () => {
+      //let triggerSpy, functionSpy;
+
+      //beforeEach(() => {
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:198]: issue: please remove commented out code

File: packages/@webex/plugin-meetings/src/annotation/index.ts [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+  };
+
+  /**
+   * Listen to websocket messages
+   * @returns {undefined}
+   */
+  private listenToEvents() {
+    if (!this.hasSubscribedToEvents) {
+      // @ts-ignore
+      this.webex.internal.mercury.on('event:locus', this.eventCommandProcessor);
```
Comment (Coread) [inline packages/@webex/plugin-meetings/src/annotation/index.ts:114]: issue: I can see a test checking for llm listening, but I can't find a test checking for mercury listening

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+      it('decryptContent', () => {
+
+        annotationService.decryptContent("decryptionKeyUrl", "content");
+        assert.calledOnceWithExactly(webex.internal.encryption.decryptText, "decryptionKeyUrl", "content");
+      });
+
+    });
+
+    describe('#sendStrokeData', () => {
+
```
Comment (Coread) [inline packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:134]: issue: There should be a test to check it does nothing if llm is not connected

**Review by Coread — CHANGES_REQUESTED** [review]
please see comments

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts (hunk 7 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:198]: Done

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts (hunk 9 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:134]: Done

File: packages/@webex/plugin-meetings/src/annotation/index.ts (hunk 1 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/src/annotation/index.ts:202]: Done

File: packages/@webex/plugin-meetings/src/annotation/index.ts [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+      url: `${this.approvalUrl}`,
+      body,
+    });
+  };
+
+  /**
+   * decrypt data
+   * @param {string} encryptionKeyUrl
+   * @param {string} content original content
+   * @returns {string} encrypted content
```
[the remark this answers is not in the document]
Comment (perfect) [reply packages/@webex/plugin-meetings/src/annotation/index.ts:232]: Done

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 11]
```diff
@@ -613,7 +615,15 @@ export default class Meeting extends StatelessWebexPlugin {
      * @memberof Meeting
      */
     // @ts-ignore
-    this.breakouts = new Breakouts({meetingId: this.id}, {parent: this.webex});
+    this.breakouts = new Breakouts({}, {parent: this.webex});
```
[the remark this answers is not in the document]
Comment (perfect) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:618]: Done

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts (hunk 4 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:1]: Done

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts (hunk 5 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:47]: got it

File: packages/@webex/plugin-meetings/src/annotation/index.ts (hunk 8 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/src/annotation/index.ts:114]: Add mercury listening check

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts (hunk 6 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:89]: Add new unit test for these cases

File: packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+        annotationService.decryptContent = sinon.stub().returns(Promise.resolve('decryptedContent'));
+      });
+
+      it('eventCommandProcessor', () => {
+        const spy = sinon.spy();
+
+        annotationService.on(EVENT_TRIGGERS.ANNOTATION_COMMAND, spy);
+
+        annotationService.eventCommandProcessor({
+          data: {
```
[the remark this answers is not in the document]
Comment (perfect) [reply packages/@webex/plugin-meetings/test/unit/spec/annotation/annotation.ts:58]: Add new UT for these cases

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 2 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:2285]: Done

File: packages/@webex/plugin-meetings/src/meeting/request.ts (hunk 3 above)
Comment (perfect) [reply packages/@webex/plugin-meetings/src/meeting/request.ts:695]: add new changeMeetingFloor unit test

**Review by Coread — APPROVED** [review]

**Pull request merged**